source/python/convert_intrinsics_opencv_to_halcon.py

```diff
@@ -0,0 +1,297 @@
+            data = yaml.safe_load(stream)
+        return OpenCVParameters(
+            Cx=data["CameraIntrinsics"]["CameraMatrix"]["CX"],
+            Cy=data["CameraIntrinsics"]["CameraMatrix"]["CY"],
+            Fx=data["CameraIntrinsics"]["CameraMatrix"]["FX"],
+            Fy=data["CameraIntrinsics"]["CameraMatrix"]["FY"],
+            k1=data["CameraIntrinsics"]["Distortion"]["K1"],
+            k2=data["CameraIntrinsics"]["Distortion"]["K2"],
+            k3=data["CameraIntrinsics"]["Distortion"]["K3"],
+            p1=data["CameraIntrinsics"]["Distortion"]["P1"],
+            p2=data["CameraIntrinsics"]["Distortion"]["P2"],
+        )
+
+    def _get_camera_matrix(self) -> NDArray[3, 3]:
+        K = np.eye(3)
+        K[0, 0] = self.Fx
+        K[1, 1] = self.Fy
+        K[0, 2] = self.Cx
+        K[1, 2] = self.Cy
+        return K
+
+    def _get_distortion(self) -> NDArray[5]:
+        return np.array([self.k1, self.k2, self.p1, self.p2, self.k3])
+
+    def undistorted_pixels(self, pixels: NDArray[Any, Any, 2]) -> NDArray[Any, Any, 2]:
+
+        undistorted_normalized = cv2.undistortPoints(
+            pixels, cameraMatrix=self._get_camera_matrix(), distCoeffs=self._get_distortion()
+        )
+        return undistorted_normalized * np.array([[[self.Fx, self.Fy]]]) + np.array([[[self.Cx, self.Cy]]])
+
+
+@dataclass
+class HalconParameters:
+    Cx: float
+    Cy: float
+    Sx: float
+    Sy: float
+    Focus: float
+    Poly: NDArray[float]
+    ImageSize: NDArray[int]
+
+    @staticmethod
+    def initial_halcon_parameters(
+        opencv_parameters: OpenCVParameters, pixel_size: float, image_size: NDArray[2]
+    ) -> HalconParameters:
+        return HalconParameters(
+            Cx=opencv_parameters.Cx,
+            Cy=opencv_parameters.Cy,
+            Sx=pixel_size,
+            Sy=pixel_size,
+            Focus=pixel_size * opencv_parameters.Fx,
+            Poly=np.zeros((5)),
+            ImageSize=image_size,
+        )
+
+    def undistorted_pixels(self, pixels: NDArray[Any, Any, 2]) -> NDArray[Any, Any, 2]:
+        x = (pixels[:, :, 0] - self.Cx) * self.Sx
+        y = (pixels[:, :, 1] - self.Cy) * self.Sy
+
+        r2 = x ** 2 + y ** 2
+        r4 = r2 ** 2
+        r6 = r2 * r4
+
+        x_u = (
+            x * (1 + self.Poly[0] * r2 + self.Poly[1] * r4 + self.Poly[2] * r6)
+            + self.Poly[3] * (r2 + 2 * x ** 2)
+            + 2 * self.Poly[4] * x * y
+        )
+        y_u = (
+            y * (1 + self.Poly[0] * r2 + self.Poly[1] * r4 + self.Poly[2] * r6)
+            + self.Poly[4] * (r2 + 2 * y ** 2)
+            + 2 * self.Poly[3] * x * y
+        )
+
+        undistorted_pixels = np.zeros(pixels.shape)
+        undistorted_pixels[:, :, 0] = x_u / self.Sx + self.Cx
+        undistorted_pixels[:, :, 1] = y_u / self.Sy + self.Cy
+        return undistorted_pixels
+
+    def _get_dat_text(self) -> str:
+        return "\n".join(
+            [
+                f"# INTERNAL CAMERA PARAMETERS",
```

Comment:
It would have to look like this (messy indentation) to output the same results.

```
    def _get_dat_text(self) -> str:
        return "\n".join(
            [
                f"""# INTERNAL CAMERA PARAMETERS
\nParGroup: Camera: Parameter;
  "Internal camera parameters";
\nFocus:foc:      {self.Focus:.06e};
  DOUBLE:0.0:;
  "Focal length of the lens [meter]";
\nPoly1:poly1:   {self.Poly[0]:.06e};
  DOUBLE::;
  "1st polynomial distortion coefficient";
\nPoly2:poly2:   {self.Poly[1]:.06e};
  DOUBLE::;
  "2nd polynomial distortion coefficient";
\nPoly3:poly3:   {self.Poly[2]:.06e};
  DOUBLE::;
  "3rd polynomial distortion coefficient";
\nPoly4:poly4:   {self.Poly[3]:.06e};
  DOUBLE::;
  "4th polynomial distortion coefficient";
\nPoly5:poly5:   {self.Poly[4]:.06e};
  DOUBLE::;
  "5th polynomial distortion coefficient";
\nSx:sx:   {self.Sx:.06e};
  DOUBLE:0.0:;
  "Width of a cell on the sensor";
\nSy:sy:   {self.Sx:.06e};
  DOUBLE:0.0:;
  "Height of a cell on the sensor";
\nCx:cx:   {self.Cx:.06e};
  DOUBLE::;
  "X-coordinate of the image center";
\nCy:cy:   {self.Cy:.06e};
  DOUBLE::;
  "Y-coordinate of the image center";
\nImageWidth:imgw:  {self.ImageSize[0]};
  INT:1:32768;
  "Width of the images";
\nImageHeight:imgh:  {self.ImageSize[1]};
  INT:1:32768;
  "Height of the images";
"""
            ]
        )
```